src/Featurizers/Components/WindowFeaturizerBase.h

```diff
@@ -0,0 +1,244 @@
+        // a vector so the "start" will change based on how much data has been overwritten.
+        size_t                                  _start_offset;
+
+        // Current number of element held by this container. It start from 0 and grows to _max_size
+        size_t                                  _cur_size;
+
+        // The vector that holds the actual data. This is set to hold _max_size elements and then
+        // is never resized.
+        std::vector<T>                          _data;
+
+    public:
+        // ----------------------------------------------------------------------
+        // |  Public Data
+        typedef CircularIterator<T>             iterator;
+        typedef CircularIterator<const T>       const_iterator;
+
+        // ----------------------------------------------------------------------
+        // |  Public Methods
+        CircularBuffer(size_t max_size) : _max_size(max_size), _start_offset(0),  _cur_size(0) {
+            _data.reserve(_max_size);
+        }
+
+        iterator begin() {
+            // When _cur_size is 0 we want to return an "end" iterator.
+            if (_cur_size == 0) {
+                return iterator();
+            }
+            return iterator(&_data[0], _max_size, _cur_size, _start_offset);
+        }
+
+        const_iterator begin() const {
+            return begin();
+        }
+
+        const_iterator cbegin() const {
+            return begin();
+        }
+
+        iterator end() {
+            // End location of the iterator is equal to the start_offset if the buffer is full,
+            // or equal to the _cur_size if it isnt full.
+            if (_cur_size == 0) {
+                return iterator();
+            }
+            return iterator(&_data[0], _max_size, 0, _cur_size == _max_size ? _start_offset : _cur_size);
+        }
+
+        const_iterator end() const {
+            return end();
+        }
+
+        const_iterator cend() const {
+            return end();
+        }
+
+        void push (T value) {
+            // Make sure we update size and offset accordingly and insert correctly.
+            // If we have filled the whole vector, then just update offset.
+            // If vector still has space, then just update the _cur_size;
+            if (_cur_size == _max_size) {
+                _data[_start_offset] = value;
+                ++_start_offset;
+                _start_offset %= _max_size;
+            } else {
+                _data.emplace_back(std::move(value));
+                ++_cur_size;
+            }
+        }
+
+        // provide a pair of begin and end iterator for the n elements requested
+        std::tuple<iterator, iterator> range(size_t n, size_t offset=0) {
+            // Since this class is used for window operations only
+            // number of requested elements is never bigger than the max_size
+            assert(n + offset <= _max_size);
```

Comment:
>assert(n + offset <= _max_size); [](start = 12, length = 32)

This assert will only be active for debug builds. This should be an exception instead and there should be a test that exercises the functionality.

Reply:
Good to know. We used an assert because we tested with a normal std::iterator and it uses an assert, but if the assert is only active for debug then its not as helpful. We did verify it works, but since we can't "catch" an assert in the test we had to remove the test for the build. Ill make the change and add the test back in.

Reply:
Per our offline discussion. Leaving as asserts for now.